includes/Core/Notifications/Notifications.php

```diff
@@ -0,0 +1,262 @@
+										'notification_id' => $data['notificationID'],
+										'notification_state' => $data['notificationState'],
+									),
+								)
+							);
+
+							if ( is_wp_error( $response ) ) {
+								return $response;
+							}
+
+							try {
+								$response = $this->parse_response( $response );
+							} catch ( Exception $e ) {
+								return new WP_Error( 'exception', $e->getMessage() );
+							}
+
+							return new WP_REST_Response(
+								array(
+									'success' => isset( $response['success'] ) ? (bool) $response['success'] : false,
+								)
+							);
+						},
+						'args'                => array(
+							'data' => array(
+								'required' => true,
+								'type'     => 'object',
+							),
+						),
+						'permission_callback' => $can_setup,
+					),
+				)
+			),
+		);
+	}
+
+	/**
+	 * Validates and parses the given JSON response into an array.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $response HTTP response array.
+	 * @return mixed JSON decoded response.
+	 * @throws Exception Throws exception if response cannot be parsed or if an error is returned.
+	 */
+	private function parse_response( $response ) {
+		$body    = wp_remote_retrieve_body( $response );
+		$decoded = json_decode( $body, true );
+
+		if ( json_last_error() ) {
+			throw new Exception( 'Error while decoding response: ' . json_last_error() );
+		}
+
+		if ( ! empty( $decoded['error'] ) ) {
+			throw new Exception( $decoded['error'] );
+		}
+
+		return $decoded;
+	}
+
+	/**
+	 * Maps the response objects into Notification objects.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $response Array of notification objects from API.
+	 * @return Notification[] Array of Notification objects.
+	 */
+	private function map_response_to_notifications( array $response ) {
+		return array_map(
+			function ( $notification ) {
+				return new Notification(
+					$notification['id'],
+					array(
+						'title'            => $notification['title'],
+						'content'          => $notification['description'],
+						'cta_url'          => $notification['ctaURL'],
+						'cta_label'        => $notification['ctaLabel'],
+						'cta_target'       => $notification['ctaTarget'],
+						'learn_more_url'   => $notification['learnMoreURL'],
+						'learn_more_label' => $notification['learnMoreLabel'],
+						'dismissible'      => $notification['dismissible'],
+						'dismiss_label'    => $notification['dismissLabel'],
+					)
+				);
+			},
+			$response
+		);
+	}
+
+	/**
+	 * Gets a WP_Error instance for the given missing required parameter.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $param Missing required parameter.
+	 * @return WP_Error
+	 */
+	private function missing_required_param( $param ) {
+		return new WP_Error(
+			'missing_required_param',
+			/* translators: %s: Missing parameter name */
+			sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), $param ),
+			array( 'status' => 400 )
+		);
+	}
```

Comment:
@felixarntz I've thought about refactoring with something like this before since there is quite a bit of duplication for this kind of thing in the datapoint handlers. Ideally we would use core's `args` definitions and leverage the native validation the API already provides. The problem is that our data requests stuff all the parameters into one `data` parameter (except for a few outliers like module activation), and the REST API doesn't support nested validation except for collections. This may not be the best place to discuss this but just thought I'd raise it.

Reply:
Yeah I thought about this initially too, but our support for batch requests would make use of that core feature challenging. I think for now we should keep this internal to the respective handlers (although adding utility methods, potentially even a validation class or trait for common validation features is certainly worth considering).

Reply:
> Yeah I thought about this initially too, but our support for batch requests would make use of that core feature challenging.

It would require that we change the way batch requests are handled to dispatch the individual REST requests (internally-w/o HTTP) in the batch handler rather than the way it is now but I think it's possible. Better to discuss this in a new issue though.